improve postgres extended query protocol troubleshooting (#14032)
DataDog/documentation: content/en/database_monitoring/setup_postgres/troubleshooting.md

| The Agent is not able to execute a required function in the `datadog` schema of the database. | The Agent requires the function `datadog.explain_statement(...)` to exist in **all databases** the Agent can collect queries from. Ensure this function was created by the root user according to the [setup instructions][1] and that the `datadog` user has permission to execute it.                                                                                                                                                                                    |
| Queries are truncated. | See the section on [truncated query samples](#query-samples-are-truncated) for instructions on how to increase the size of sample query text.                                                                                                                                                                                                                                                                                                                              |

> | The application client used to execute the query is using the Postgres extended query protocol or prepared statements. | See the section on [Postgres extended query protocol](#postgres-extended-query-protocol).                                                                                                                                                                                                                                                                                                                                                                                  |

Reviewer: ```suggestion
| The application client used to execute the query is using the Postgres extended query protocol or prepared statements. | See [Postgres extended query protocol](#postgres-extended-query-protocol).                                                                                                                                                                                                                                                                                                                                                                                  |
```

SELECT DISTINCT address FROM customers WHERE id = ANY(ARRAY[ ?
```

To avoid this, raise the `track_activity_query_size` setting to a value large enough to accommodate the largest expected text size of your queries. For further information, see the Postgres documentation on [runtime statistics][12].

### Queries are missing explain plans

Some or all queries may not have plans available. This can be due to unsupported query commands, queries made by unsupported client applications, an outdated Agent, or incomplete database setup. Below are possible causes for missing explain plans.

#### Missing explain function {#undefined-explain-function}

> Problem: The Agent is not able to execute a required function in the `datadog` schema of the database.

Reviewer: ```suggestion
Problem: The Agent is unable to execute a required function in the `datadog` schema of the database.
```

> Solution: The Agent requires the function `datadog.explain_statement(...)` to exist in **all databases** the Agent can collect queries from.

Reviewer: ```suggestion
Solution: The Agent requires the `datadog.explain_statement(...)` function to exist in **all databases** the Agent can collect queries from.
```

#### Queries are truncated
See the section on [truncated query samples](#query-samples-are-truncated) for instructions on how to increase the size of sample query text.

#### Postgres extended query protocol

> If a client is using the Postgres [extended query protocol][13] or prepared statements then the Datadog Agent is not able to collect explain plans due to the separation of the parsed query and raw bind parameters. If the client provides an option to force using the simple query protocol then turning that on will enable the Datadog Agent to collect execution plans.

Reviewer: ```suggestion
If a client is using the Postgres [extended query protocol][13] or prepared statements, the Datadog Agent is unable to collect explain plans due to the separation of the parsed query and raw bind parameters. If the client provides an option to force using the simple query protocol, then turning that on enables the Datadog Agent to collect execution plans.
```

| Language | Client                    | Configuration for simple query protocol                                                                                                                                                                                                                                                                                                                                                      |
|----------|---------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | Go       | [pgx][6]                  | Set `PreferSimpleProtocol` to switch to the simple query protocol.                                                                                                                                                                                                                                                                                                                           |

Reviewer 2: ```suggestion
| Go       | [pgx][6]                  | Set `PreferSimpleProtocol` to switch to the simple query protocol ([Docs](https://pkg.go.dev/github.com/jackc/pgx#ConnConfig)).                                                                                                                                                                                                                                                                                                                           |
```

> | Go       | [pgx][6]                  | Set `PreferSimpleProtocol` to switch to the simple query protocol ([Docs][14]).                                                                                                                                                                                                                                                                                                              |

Reviewer: ```suggestion
| Go       | [pgx][6]                  | Set `PreferSimpleProtocol` to switch to the simple query protocol (See the [ConnConfig documentation][14]).                                                                                                                                                                                                                                                                                                              |
```

> | Java     | [Postgres JDBC Client][14] | Set `preferQueryMode = simple` to switch to the simple query protocol.                                                                                                                                                                                                                                                                                                                       |

Reviewer 2: ```suggestion
| Java     | [Postgres JDBC Client][14] | Set `preferQueryMode = simple` to switch to the simple query protocol ([Docs](https://jdbc.postgresql.org/documentation/publicapi/org/postgresql/jdbc/PreferQueryMode.html)).                                                                                                                                                                                                                                                                                                                       |
```

> | Java     | [Postgres JDBC Client][15] | Set `preferQueryMode = simple` to switch to the simple query protocol ([Docs][16]).                                                                                                                                                                                                                                                                                                          |

Reviewer: ```suggestion
| Java     | [Postgres JDBC Client][15] | Set `preferQueryMode = simple` to switch to the simple query protocol (See the [PreferQueryMode documentation][16]).                                                                                                                                                                                                                                                                                                          |
```

> | Python   | [asyncpg][17]              | Uses the extended query protocol and it's not possible to disable it. Disabling prepared statements does not solve the problem.                                                                                                                                                                                                                                                              |

Reviewer 3: Is there a reason why we don't have the pre-formatted statement workaround for `asyncpg` like we have for the others below? 
```
Format SQL Queries using [psycopg sql][17] before passing them to the DB client to enable collection of execution plans.
```

Author: Good question. I couldn't find a built-in sql formatter in `asyncpg`. We should still present a workaround though. I added some wording referring to the `psycopg` formatter.

> | Python   | [asyncpg][17]              | Uses the extended query protocol and it's not possible to disable it. Disabling prepared statements does not solve the problem.                                                                                                                                                                                                                                                              |

Reviewer: ```suggestion
| Python   | [asyncpg][17]              | Uses the extended query protocol, which cannot be disabled. Disabling prepared statements does not solve the problem.                                                                                                                                                                                                                                                              |
```

> | Python   | [psycopg][18]             | `psycopg2` does not use the extended query protocol so execution plans should be collected without issue. <br/> `psycopg3` uses the extended query protocol by default and there is no way to disable it. Disabling prepared statements does not solve the problem. Format SQL Queries using [psycopg sql][19] before passing them to the DB client to enable collection of execution plans. |

Reviewer: ```suggestion
| Python   | [psycopg][18]             | `psycopg2` does not use the extended query protocol so execution plans should be collected without issue. <br/> `psycopg3` uses the extended query protocol by default and cannot be disabled. Disabling prepared statements does not solve the problem. To enable the collection of execution plans, format SQL Queries using [psycopg sql][19] before passing them to the DB client. |
```

> | Node     | [node-postgres][20]       | Uses the extended query protocol and it's not possible to disable it. Use [pg-format][21] to format SQL Queries before passing them to [node-postgres][20] to enable the Datadog Agent to collect execution plans.                                                                                                                                                                           |

Reviewer: ```suggestion
| Node     | [node-postgres][20]       | Uses the extended query protocol and cannot be disabled. To enable the Datadog Agent to collect execution plans, use [pg-format][21] to format SQL Queries before passing them to [node-postgres][20].                                                                                                                                                                           |
```

sudo apt-get install postgresql-contrib-10
```

> See the appropriate version of the [Postgres `contrib` documentation][22] for more information.

Reviewer: ```suggestion
For more information, see the appropriate version of the [Postgres `contrib` documentation][22].
```
